### PR TITLE
Scale the tile size limit up with the multiplier at low zooms

### PR DESCRIPTION
**Pull request by e-n-f**
Prior to this PR, the tile size limit was the same at every zoom level. With this PR, if you specify a `--retain-points-multiplier`, the tile size limit will be scaled up at low zooms that are expected to contain more features because of the multiplier, and will remain at its usual level at basezoom and higher, where there are no extra features for the multiplier to retain.

This PR also contains a fix for the "compression error 106" problem, which turns out to have happened when bailing out after a tile couldn't be made small enough. In this case it now goes through the standard post-tile-generation cleanup instead of taking shortcuts.

It also contains a change that makes it less likely that a tile can't be made small enough, by trying harder to find a suitable size threshold if the `--drop-smallest-as-needed` guess yields the current threshold again, which can happen if there are large numbers of points at exactly the same location, which are treated as having an area of 0.

### CODE REVIEW

File: tile.cpp [hunk 1]
```diff
@@ -361,7 +361,7 @@ std::vector<std::vector<coalesce>> assemble_multiplier_clusters(std::vector<coal
 				}
 			}
 
-			if (is_cluster_start) {
+			if (is_cluster_start || clusters.size() == 0) {
```
Comment (e-n-f) [inline tile.cpp:364]: Fixes segfault

File: tile.cpp [hunk 2]
```diff
@@ -1358,9 +1358,15 @@ unsigned long long choose_mingap(std::vector<unsigned long long> const &indices,
...
+long long choose_minextent(std::vector<long long> &extents, double f, long long existing_extent) {
 	std::sort(extents.begin(), extents.end());
-	return extents[(extents.size() - 1) * (1 - f)];
+	size_t off = (extents.size() - 1) * (1 - f);
+
+	while (off + 1 < extents.size() && extents[off] == existing_extent) {
+		off++;
+	}
+
+	return extents[off];
```
Comment (e-n-f) [inline tile.cpp:1369]: Searches harder for a viable feature size threshold before giving up

File: tile.cpp [hunk 3]
```diff
@@ -1960,6 +1968,20 @@ long long write_tile(decompressor *geoms, std::atomic<long long> *geompos_in, ch
...
+	size_t scaled_max_tile_size = max_tile_size;
+	double regular_retention = 1 / exp(log(arg->droprate) * (arg->basezoom - z));
+	if (regular_retention > 1) {
+		regular_retention = 1;
+	}
+	double multiplier_retention = 1 / exp(log(arg->droprate) * (arg->basezoom - z)) * retain_points_multiplier;
+	if (multiplier_retention > 1) {
+		multiplier_retention = 1;
+	}
+	scaled_max_tile_size *= multiplier_retention / regular_retention;
```
Comment (e-n-f) [inline tile.cpp:1983]: Increases tile size at low zooms according to multiplier

File: tile.cpp [hunk 4]
```diff
@@ -3090,6 +3107,12 @@ void *run_thread(void *vargs) {
 				perror("pthread_mutex_unlock");
 				exit(EXIT_PTHREAD);
 			}
+
+			if (len < 0) {
+				err_or_null = &arg->err;
+				*err_or_null = z - 1;
+				break;
```
Comment (e-n-f) [inline tile.cpp:3114]: Delays early exit until after cleanup

File: man/tippecanoe.1 [hunk 5]
```diff
@@ -29,7 +29,11 @@ significant travel corridor.
 .PP
 The easiest way to install tippecanoe on OSX is with Homebrew \[la]http://brew.sh/\[ra]:
 .PP
+.RS
```
Comment (ChrisLoer) [inline man/tippecanoe.1:32]: What does this notation mean?

File: tile.cpp [hunk 6]
```diff
@@ -1358,9 +1358,15 @@ unsigned long long choose_mingap(std::vector<unsigned long long> const &indices,
 	return top;
 }
 
-long long choose_minextent(std::vector<long long> &extents, double f) {
```
Comment (ChrisLoer) [inline tile.cpp:1361]: A comment explaining the logic here would be nice -- specifically what are the "extents" and what's the "fraction" represent conceptually.
Is `off` just short for `offset`? If so spelling it out seems a bit more legible

**Review by ChrisLoer — COMMENTED** [review]
I haven't tested yet but the basics of the changes all make sense to me

**Review by ChrisLoer — APPROVED** [review]
OK, I got a chance to test in Felt -- my tests were using the same files you ran, so maybe limited new information, but it behaved as expected.

File: man/tippecanoe.1 (hunk 5 above)
Comment (e-n-f) [reply man/tippecanoe.1:32]: This is `nroff -man` format, the ancient markup language of manpages. I haven't written one by hand for so long that I can't remember what `.RS` does (this one is generated by `md2man-roff`), although I used to care an awful lot about nroff and troff, but:

```
$ man 7 man
MAN(7)                 Miscellaneous Information Manual                 MAN(7)

NAME
     man – legacy formatting language for manual pages
…
     RS, RE    reset the left margin: [width]
```

File: tile.cpp (hunk 6 above)
Comment (e-n-f) [reply tile.cpp:1361]: Thanks, I'll add a comment.

The "extents" are the areas of polygons or the pseudo-areas associated with linestrings and points. The "fraction" is the proportion of the features that tippecanoe is trying to preserve in the hope of making the tile small enough, so when it says

```
Going to try keeping the biggest 20.19% of the features to make it fit
```

the fraction is 0.2019. And yes, "off" is the array index of the "extent" that it hopes will be good enough to make the tile the right size.

Comment (e-n-f) [reply tile.cpp:1361]: Comment added

**Pull request merged**